### PR TITLE
[Kernels] Overlap shared experts with send/recv

### PR DESCRIPTION
**Pull request by bnellnm**
## Purpose

Overlap shared experts computation with send and receive operations in all2all dispatcher.

TODO
- Implement `prepare_async` for flashinfer prepare/finalize class?
- Move sum of shared/fused experts and final reduce of into the FusedMoE layer.

## Test Plan
Tried it with llama4 and deepseek_v2
Added unit test for shared experts in `test_pplx_moe.py`

## Test Result

This trace shows the MLP (nvjet...) overlapped with DeepEP dispatch send/receive (green).

Screenshot 2025-08-29 at 3 19 55 PM

## Documentation Update
Updated modular kernel docs

cc @SageMoore , @LucasWilkinson

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py [hunk 1]
```diff
@@ -84,12 +84,15 @@ def max_num_tokens_per_rank(self) -> Optional[int]:
...
 
     def topk_indices_dtype(self) -> Optional[torch.dtype]:
-        return torch.int32
+        return torch.uint32
 
     def num_dispatchers(self) -> int:
         return self.num_dispatchers_
 
-    def prepare(
+    def has_prepare_no_receive(self) -> bool:
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py]: nit: I think a better names might be 

`has_prepare_no_receive` -> `supports_async`

and:

`prepare_no_receive` -> `prepare_and_send_async`

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py]: What about just `prepare_async`?

File: vllm/model_executor/layers/quantization/deepgemm.py [hunk 2]
```diff
@@ -23,6 +23,8 @@ def prepare_block_fp8_matmul_inputs(
     assert len(block_size) == 2
     block_n, block_k = block_size[0], block_size[1]
 
+    print(f"A/B {A.shape} {B.shape}")
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/deepgemm.py]: nit: cruft

File: vllm/model_executor/models/llama4.py [hunk 3]
```diff
@@ -73,35 +74,70 @@ def __init__(self,
...
-            intermediate_size=intermediate_size_moe,
-            apply_router_weight_on_input=True,
-            reduce_results=False,
-            renormalize=False,
-            quant_config=quant_config,
-            prefix=f"{prefix}.experts")
+        # This is a temporary flag for testing.
+        self.use_shared_fused = True
+
+        if not self.use_shared_fused:
```
Comment (LucasWilkinson) [inline vllm/model_executor/models/llama4.py]: is there any reason to not `use_shared_fused`? since for non-async backends it just degenerates to effectively the same thing? I guess the only downside is no torch.compile inside the FusedMoE?

File: vllm/model_executor/models/deepseek_v2.py [hunk 4]
```diff
@@ -146,61 +147,110 @@ def __init__(
...
+                final_hidden_states = self.experts(hidden_states=hidden_states,
+                                                   router_logits=router_logits)
+            if shared_output is not None:
+                if hidden_states.dtype != torch.float16:
+                    final_hidden_states = final_hidden_states + shared_output
+                else:
+                    # Fix FP16 overflow
+                    # See DeepseekV2DecoderLayer for more details.
+                    final_hidden_states = final_hidden_states + shared_output \
+                        * (1. / self.routed_scaling_factor)
```
Comment (LucasWilkinson) [inline vllm/model_executor/models/deepseek_v2.py]: would if be easier if we collapsed alot of this to something like
```
fused_moe_out  = self.experts(hidden_states=hidden_states, router_logits=router_logits)

if self.shared_experts is not None:
     shared_output = self.shared_experts(hidden_states)
     routed_output = fused_moe_out
else:
     shared_output, routed_output = fused_moe_out

if hidden_states.dtype != torch.float16:
     routed_output *= self.routed_scaling_factor
     shared_output *= shared_output * (1. / self.routed_scaling_factor)

final_hidden_states = routed_output + shared_output
```

File: vllm/model_executor/models/llama4.py (hunk 3 above)
Comment (bnellnm) [reply vllm/model_executor/models/llama4.py]: I left it as a flag mostly for debugging.  I am a bit concerned about the effects of torch.compile.  Maybe we can wrap some of the innards of the modular kernel with torch.compile to get that back.  I'm fine with making it on by default or adding an env var, etc.

I'm trying to do some benchmarks now to see how it is working.

File: vllm/model_executor/models/deepseek_v2.py (hunk 4 above)
Comment (bnellnm) [reply vllm/model_executor/models/deepseek_v2.py]: @LucasWilkinson can you double check this bit?   I refactored it in a different way that preserves the original execution order.  Based on the original code, I think that the scaling of `shared_output` only happens when the type `== float16` as opposed to `!= float16`.

File: vllm/model_executor/models/llama4.py (hunk 3 above)
Comment (SageMoore) [reply vllm/model_executor/models/llama4.py]: I'm somewhat against having this flag. IMO this should just always be on.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 5]
```diff
@@ -1700,14 +1756,30 @@ def forward_impl(self, hidden_states: torch.Tensor,
             logical_replica_count=self.logical_replica_count,
         )
 
-        if do_naive_dispatch_combine:
-            final_hidden_states = get_ep_group().combine(final_hidden_states)
-        if self.reduce_results and (self.tp_size > 1 or self.ep_size > 1):
-            # Default set to False. (May have to add shared expert outputs.
-            final_hidden_states = self.maybe_all_reduce_tensor_model_parallel(
-                final_hidden_states)
+        if shared_output is not None:
```
Comment (SageMoore) [inline vllm/model_executor/layers/fused_moe/layer.py:1782]: Nit: `# if the shared_output was computed, add it to the final_hidden_states`

Furthermore, I think it may be a bit safer to reconfigure this check to
```
if self.shared_experts is not None:
    assert shared_output is not None
...
```

That way we keep the check consistent throughout the function and keep ourselves safe if we are expecting a shared_output but we don't get one back from the modular kernel?

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 6]
```diff
@@ -141,6 +141,17 @@ def apply(self, output: Optional[torch.Tensor],
         raise NotImplementedError
 
 
+PrepareResultType = tuple[
```
Comment (SageMoore) [inline vllm/model_executor/layers/fused_moe/modular_kernel.py:156]: I think a `@dataclass` would be good here. If you prefer the tuple, let's just put comments on what each of the items represent.

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 7]
```diff
@@ -795,4 +874,7 @@ def forward(
             self.fused_experts.finalize_weight_and_reduce_impl(),
         )
 
-        return output
+        if self.shared_experts is None:
+            return output
+        else:
```
Comment (SageMoore) [inline vllm/model_executor/layers/fused_moe/modular_kernel.py:891]: Nit: `assert shared_output is not None`?

**Review by SageMoore — COMMENTED** [review]
Nice work, @bnellnm. I only have minor nits. Otherwise looks good.

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 7 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:891]: `shared_output` is declared to be a `torch.Tensor` so it should never be `None`

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 5 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py:1759]: Unfortunately, self.shared_experts always needs to be set even if we aren't doing the overlap due to weight loading/processing.  I can use `config.n_shared_experts is not None` instead.

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 6 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:156]: I'll add some comments.

**Review by LucasWilkinson — APPROVED** [review]
LGTM! Would be good to get a trace though to show the overlap

**Review by SageMoore — APPROVED** [review]

File: docs/design/fused_moe_modular_kernel.md [hunk 8]
```diff
@@ -54,8 +54,8 @@ The `FusedMoEModularKernel` acts as a bridge between the `FusedMoEPermuteExperts
...
+The `FusedMoEPrepareAndFinalize` abstract class exposes `prepare`, `prepare_no_receive`  and `finalize` functions.
+The `prepare` function is responsible for input activation Quantization and All2All Dispatch. If implemented, The `prepare_no_receive` is like `prepare` except it does not wait to receive results from other workers.  Instead it returns a "receiver" thunk that must be called to wait for the final results of worker. It is not required that this method is supported by all `FusedMoEPrepareAndFinalize` classes, but if it is available, it can be used to interleave work with the initial all to all communication, e.g. interleaving shared experts with fused experts.  The `finalize` function is responsible for invoking the All2All Combine. Additionally the `finalize` function may or may not do the TopK weight application and reduction (Please refer to the TopKWeightAndReduce section)
```
Comment (hmellor) [inline docs/design/fused_moe_modular_kernel.md]: thunk?

Comment (bnellnm) [reply docs/design/fused_moe_modular_kernel.md]: A parameterless function.  I can change the wording if it's not clear.

**Comment by bnellnm** [comment]
@SageMoore , @LucasWilkinson could you guys take a final look?  Should be ready to merge.  @robertgshaw2-redhat verified that it works on multi-node systems now.

**Comment by yewentao256** [comment]
This PR introduces the R1 accuracy issue https://github.com/vllm-project/vllm/issues/24530
